File: Desktop/GrappBox/Body/BugTracker/BugEntity.cpp
#include "BugEntity.h"

SBugState::SBugState(const int id, const QString &name)
{
    _id = id;
    _name = name;
}

SBugState::SBugState() : SBugState(-1, "") {}
SBugState::~SBugState() {}

//System
BugEntity::BugEntity()
{
    _id = -1;
}

BugEntity::BugEntity(QJsonObject obj)
{
    _id = obj["id"].toInt();
    _authorId = obj["creatorId"].toInt();
    _projectId = obj["projectId"].toInt();
    _title = obj["title"].toString();
    _description = obj["description"].toString();
    _parentId = obj["parentId"].toInt();
    if (!obj["createdAt"].isNull())
        _createdAt = QDateTime::fromString(obj["createdAt"].toObject()["date"].toString(), "yyyy-MM-dd hh:mm:ss");
    if (!obj["editedAt"].isNull())
        _editedAt = QDateTime::fromString(obj["editedAt"].toObject()["date"].toString(), "yyyy-MM-dd hh:mm:ss");
    if (!obj["deletedAt"].isNull())
        _deletedAt = QDateTime::fromString(obj["deletedAt"].toObject()["date"].toString(), "yyyy-MM-dd hh:mm:ss");
    _state = SBugState(obj["state"].toObject()["id"].toInt(), obj["state"].toObject()["name"].toString());
    if (!obj["tags"].isNull())
    {
        QJsonArray tags = obj["tags"].toArray();
        QJsonArray::iterator tagIt;

        for (tagIt = tags.begin(); tagIt != tags.end(); ++tagIt)
            _tags.append(BugTagEntity((*tagIt).toObject()["id"].toInt(), (*tagIt).toObject()["name"].toString()));
    }
    if (!obj["users"].isNull())

    {
        QJsonArray users = obj["users"].toArray();
        QJsonArray::iterator userIt;

        for (userIt = users.begin(); userIt != users.end(); ++userIt)
        {
            QJsonObject current = (*userIt).toObject();

            _users.append(BugUser(current["id"].toInt(), current["name"].toString(), current["email"].toString(), QByteArray::fromBase64(current["avatar"].toString().toStdString().c_str())));
        }
    }
}

//Getters
const int BugEntity::GetID() const { return _id; }
const int BugEntity::GetAuthorId() const { return _authorId; }
const int BugEntity::GetProjectId() const { return _projectId; }
const QString &BugEntity::GetTitle() const { return _title; }
const QString &BugEntity::GetDescription() const { return _description; }
const int BugEntity::GetParentId() const { return _parentId; }
const QDateTime &BugEntity::GetCreatedAt() const { return _createdAt; }
const QDateTime &BugEntity::GetEditedAt() const { return _editedAt; }
const QDateTime &BugEntity::GetDeletedAt() const { return _deletedAt; }
const SBugState &BugEntity::GetState() const { return _state; }
const QList<BugTagEntity> &BugEntity::GetTags() const { return _tags; }
const QList<BugUser> &BugEntity::GetUsers() const { return _users; }
const bool BugEntity::IsValid() const { return _id >= 0; }

//Setters
void BugEntity::SetAuthorID(const int id) { _id = id; }
void BugEntity::SetProjectID(const int id) { _projectId = id; }
void BugEntity::SetTitle(const QString &title) { _title = title; }
void BugEntity::SetDescription(const QString &desc) { _description = desc; }
void BugEntity::SetParentID(const int id) { _parentId = id; }
void BugEntity::SetState(const SBugState &state) { _state = state; }
void BugEntity::SetTags(const QList<BugTagEntity> &tags) { _tags = tags; }
void BugEntity::AddTag(const BugTagEntity &tag) { _tags.append(tag); }
void BugEntity::DelTag(QList<BugTagEntity>::const_iterator tagIt) { _tags.removeOne(*tagIt); }
void BugEntity::SetUsers(const QList<BugUser> &users) { _users = users; }
void BugEntity::AddUser(const BugUser &user) { _users.append(user); }
void BugEntity::DelUser(QList<BugUser>::const_iterator userIt) { _users.removeOne(*userIt); }

File: Android-Work/app/src/main/java/com/grappbox/grappbox/grappbox/BugTracker/BugEntity.java
package com.grappbox.grappbox.grappbox.BugTracker;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by wieser_m on 18/02/2016.
 */
public class BugEntity {
    public static final String EXTRA_GRAPPBOX_BUG_ID = "extra.grappbox.bugId";

    String              _id;
    String              _projectId;
    String              _title;
    String              _description;
    String              _parentId;
    String              _createdAt;
    String              _editedAt;
    String              _deletedAt;
    String              _creatorId;
    String              _creatorFullname;
    String              _stateId;
    String              _stateName;
    List<TagEntity>     _tags;
    List<UserEntity>    _users;


    public BugEntity()
    {
        _id = "";
        _tags = new ArrayList<>();
        _users = new ArrayList<>();
    }

    public BugEntity(JSONObject data) throws JSONException {
        _tags = new ArrayList<>();
        _users = new ArrayList<>();
        reimport(data);
    }

    public boolean IsValid()
    {
        return !_id.isEmpty();
    }

    public void reimport(JSONObject data) throws JSONException
    {
        JSONObject creator = data.getJSONObject("creator");
        JSONObject createdAt = data.getJSONObject("createdAt");
        JSONObject editedAt = (data.has("editedAt") && !data.isNull("editedAt") ? data.getJSONObject("editedAt") : null);
        JSONObject deletedAt = (data.has("deletedAt") && !data.isNull("deletedAt") ? data.getJSONObject("deletedAt") : null);
        JSONArray tags = data.getJSONArray("tags");
        JSONArray users = data.getJSONArray("users");
        int max = (tags.length() > users.length() ? tags.length() : users.length());

        _id = data.getString("id");
        _projectId = data.getString("projectId");
        _title = data.getString("title");
        _description = data.getString("description");
        _parentId = data.getString("parentId");
        _createdAt = createdAt.getString("date");
        _editedAt = (editedAt == null ? null : editedAt.getString("date"));
        _deletedAt = (deletedAt == null ? null : deletedAt.getString("date"));
        _creatorId = creator.getString("id");
        _creatorFullname = creator.getString("fullname");

        for (int i = 0; i < max; ++i)
        {
            if (i < tags.length())
                _tags.add(new TagEntity(tags.getJSONObject(i)));
            if (i < users.length())
                _users.add(new UserEntity(users.getJSONObject(i)));
        }
    }

    public String GetId() {
        return _id;
    }

    public void SetId(String _id) {
        this._id = _id;
    }

    public String GetProjectId() {
        return _projectId;
    }

    public void SetProjectId(String _projectId) {
        this._projectId = _projectId;
    }

    public String GetTitle() {
        return _title;
    }

    public void SetTitle(String _title) {
        this._title = _title;
    }

    public String GetDescription() {
        return _description;
    }

    public void SetDescription(String _description) {
        this._description = _description;
    }

    public String GetParentId() {
        return _parentId;
    }

    public void SetParentId(String _parentId) {
        this._parentId = _parentId;
    }

    public String GetCreatedAt() {
        return _createdAt;
    }

    public void SetCreatedAt(String _createdAt) {
        this._createdAt = _createdAt;
    }

    public String GetEditedAt() {
        return _editedAt;
    }

    public void SetEditedAt(String _editedAt) {
        this._editedAt = _editedAt;
    }

    public String GetDeletedAt() {
        return _deletedAt;
    }

    public void SetDeletedAt(String _deletedAt) {
        this._deletedAt = _deletedAt;
    }

    public String GetCreatorId() {
        return _creatorId;
    }

    public void SetCreatorId(String _creatorId) {
        this._creatorId = _creatorId;
    }

    public String GetCreatorFullname() {
        return _creatorFullname;
    }

    public void SetCreatorFullname(String _creatorFullname) {
        this._creatorFullname = _creatorFullname;
    }

    public String GetStateId() {
        return _stateId;
    }

    public void SetStateId(String _stateId) {
        this._stateId = _stateId;
    }

    public String GetStateName() {
        return _stateName;
    }

    public void SetStateName(String _stateName) {
        this._stateName = _stateName;
    }

    public List<TagEntity> GetTags() {
        return _tags;
    }

    public void GetTags(List<TagEntity> _tags) {
        this._tags = _tags;
    }

    public List<UserEntity> GetUsers() {
        return _users;
    }

    public void SetUsers(List<UserEntity> _users) {
        this._users = _users;
    }

    public boolean IsClosed() {
        return !(_deletedAt == null || _deletedAt.isEmpty());
    }
}
